### PR TITLE
fqdn: Exit go routines early if datapath update times out

### PR DESCRIPTION
**Pull request by gandro**
This PR optimizes the time FQDN Go routines linger around when waiting on datapath updates. Before this PR, they would wait until the datapath update finished, even though the parent already decided to moved on without waiting on the datpath update.

See commit messages and https://github.com/cilium/cilium/pull/32769#discussion_r1626356945 for further details

### CODE REVIEW

**Review by nebril — APPROVED** [review]

File: pkg/ipcache/metadata.go [hunk 1]
```diff
@@ -158,12 +158,28 @@ func (m *metadata) setInjectedRevision(rev uint64) {
...
-func (m *metadata) waitForRevision(rev uint64) {
+func (m *metadata) waitForRevision(ctx context.Context, rev uint64) error {
+	// Allow callers to bail out by cancelling the context
+	cleanupCancellation := context.AfterFunc(ctx, func() {
+		// We need to acquire injectedRevisionCond.L here to be sure that the
+		// Broadcast won't occur before the call to Wait, which would result
+		// in a missed signal.
+		m.injectedRevisionCond.L.Lock()
+		m.injectedRevisionCond.Broadcast()
+		m.injectedRevisionCond.L.Unlock()
```
Comment (doniacld) [inline pkg/ipcache/metadata.go]: Would it make sense to defer the `Unlock` here?

Comment (gandro) [reply pkg/ipcache/metadata.go]: With these short functions, it's a matter of personal preference I would argue. I can add a defer!

File: pkg/fqdn/name_manager.go [hunk 2]
```diff
@@ -268,8 +268,9 @@ func (n *NameManager) UpdateGenerateDNS(ctx context.Context, lookupTime time.Tim
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
-		n.config.IPCache.WaitForRevision(ipcacheRevision)
-		wg.Done()
+		if err := n.config.IPCache.WaitForRevision(ctx, ipcacheRevision); err != nil {
```
Comment (gandro) [inline pkg/fqdn/name_manager.go]: As @doniacld pointed out offline, this should be `err == nil`. Will fix!

File: pkg/ipcache/metadata.go (hunk 1 above)
Comment (doniacld) [reply pkg/ipcache/metadata.go]: Follow your style, I was just wondering if it wasn't safe!

File: pkg/ipcache/metadata_test.go [hunk 3]
```diff
@@ -782,18 +783,25 @@ func TestMetadataRevision(t *testing.T) {
 func TestMetadataWaitForRevision(t *testing.T) {
 	m := newMetadata()
 
-	p1 := netip.MustParsePrefix("1.1.1.1/32")
-	wantRev := m.enqueuePrefixUpdates(p1)
+	_, wantRev := m.dequeuePrefixUpdates()
```
Comment (doniacld) [inline pkg/ipcache/metadata_test.go:786]: What is the first value that we ignore? 
When I check the signature there are only the modifiedPrefixes which are returned.

Comment (gandro) [reply pkg/ipcache/metadata_test.go:786]: Yes - but we haven't modified any prefixes, so we know this is empty and thus can ignore it. We're basically only using `dequeuePrefixUpdates`  to obtain the next revision, as `dequeuePrefixUpdates` always bumps the revision if called - even if there's nothing queued

Comment (doniacld) [reply pkg/ipcache/metadata_test.go:786]: I was checking the signature of the function on a older branch, my bad!

**Review by doniacld — APPROVED** [review]
Congrats 🙌 Good job!

**Pull request merged**